### PR TITLE
Feature/pre test refactoring

### PR DESCRIPTION
**Pull request by jenniferlynparsons**

### CODE REVIEW

File: client/src/actions/authActions.js [hunk 1]
```diff
@@ -1,100 +1,49 @@
...
+    });
   };
 }
 
-// Set logged in user
+// Log user out
+export const logoutUser = () => dispatch => {
+  localStorage.removeItem("jwtToken");
+  setAuthToken(false);
+  dispatch(resetStateOnLogout());
```
Comment (bdj61383) [inline client/src/actions/authActions.js]: Just curious why we're calling `resetStateOnLogout` instead of just writing this line as `dispatch({type: "USER_LOGOUT"})`.  Is `resetStateOnLogout` going to have side effects in the future?  If you're going to keep the function, consider renaming it, as an action creator should really describe the event/action that is occurring, not the effects of the action.  Leave that to the reducer.

File: client/src/components/Tea/TeaDetails/TeaDetails.jsx [hunk 2]
```diff
@@ -2,37 +2,40 @@
...
-        to={"/update-tea/" + props.tea.id}
-        className="button"
-        onClick={e => props.onClick(e, false)}
-      >
-        Edit
-      </Link>
-    </div>
-  );
-};
+export class TeaDetails extends React.Component {
```
Comment (bdj61383) [inline client/src/components/Tea/TeaDetails/TeaDetails.jsx]: Why the change from a pure function to a component?  For ease of testing?  If you're going to make this switch, use `React. PureComponent` to benefit from its optimizations.  Even better, consider wrapping the original function in `React.Memo`.

File: client/src/components/Tea/TeaDetails/TeaDetailsContainer.jsx [hunk 3]
```diff
@@ -2,25 +2,25 @@
...
 import { connect } from "react-redux";
-import { deleteTea, getTeas } from "../../../actions/teaActions";
+import { getTeas } from "../../../actions/teaActions";
 import { TeaDetails } from "./TeaDetails";
 import { editTeaFlash } from "../../../actions/flashActions";
 
 class TeaDetailsContainer extends Component {
-  clickHandler = (e, status) => {
-    this.props.updateFlash(status);
+  clickHandler = status => {
```
Comment (bdj61383) [inline client/src/components/Tea/TeaDetails/TeaDetailsContainer.jsx]: Up to you, but this could be a one-liner: `clickHandler = status => this.props.editTeaFlash(status)`.  One difference if you choose to make this a one-liner is that now `clickHandler` will return the result of `editTeaFlash`.  You're not doing anything with the result of `clickHandler` so it shouldn't matter, but something to keep in mind.

File: client/src/components/Tea/TeaEditor/TeaEditorContainer.jsx [hunk 4]
```diff
@@ -120,7 +116,7 @@ export class TeaEditorContainer extends React.Component {
   };
 
   componentDidMount() {
-    this.props.getTeaList(this.props.userID);
+    this.props.getTeas(this.props.userID);
   }
 
   componentDidUpdate(prevProps) {
```
Comment (bdj61383) [inline client/src/components/Tea/TeaEditor/TeaEditorContainer.jsx:111]: You also need to handle the case here where you have `this.props.currentTea` but there is no `prevProps.currentTea` (because you reloaded the page while you're on it).  Something like:
```
( this.props.currentTea && !prevProps.currentTea ) || 
( this.props.currentTea && this.props.currentTea.id !== prevProps.currentTea.id) )
```

File: client/src/components/Tea/TeaEditor/TeaEditor.jsx [hunk 5]
```diff
@@ -3,123 +3,125 @@ import React from "react";
...
-    <div className="container">
-      {props.flash.name ? (
-        <div className="notification is-success">
-          {props.flash.name} has been succesfully saved.{" "}
-          <Link to={"../../tea/" + props.flash.teaID}>View details</Link>
-        </div>
-      ) : (
-        ""
-      )}
+export class TeaEditor extends React.Component {
```
Comment (bdj61383) [inline client/src/components/Tea/TeaEditor/TeaEditor.jsx]: Same comment as for `TeaDetails` apply here.  However, I think there is a bigger issue.  Regardless of if this is a pure function or a React component, it is going to re-render every time you change one of the fields (because the props change).  Perhaps it's an over-optimization at this point, but a workaround would be to create an `Input` component, so that only that small component would re-render when its input is changed.  Something to think about down the line so that the whole form isn't re-rendered dozens or hundreds of times during its lifecycle.

File: client/src/components/Tea/TeaEditor/TeaEditor.jsx [hunk 6]
```diff
@@ -3,123 +3,125 @@ import React from "react";
...
-            </div>
-            {/* {props.shouldMarkError("name") ? (
+        <form onSubmit={this.props.handleFormSubmit}>
+          <div className="field">
+            <label htmlFor="name">
+              Tea Name
+              <div className="control">
+                <input
+                  className="input"
+                  // className={
```
Comment (bdj61383) [inline client/src/components/Tea/TeaEditor/TeaEditor.jsx]: I know you have these comments in here for when you add validation, but as a general rule of thumb, you shouldn't check in commented-out code.

**Review by bdj61383 — CHANGES_REQUESTED** [review]

File: client/src/actions/authActions.js (hunk 1 above)
Comment (jenniferlynparsons) [reply client/src/actions/authActions.js]: good question! most of the user auth stuff was grabbed from a tutorial so i hadn't refactored it other than the actions. i'll fix this up.

File: client/src/components/Tea/TeaEditor/TeaEditor.jsx (hunk 5 above)
Comment (jenniferlynparsons) [reply client/src/components/Tea/TeaEditor/TeaEditor.jsx]: yes! this is something i've been considering. i think it would also make testing a little more targeted. and it will become more of an issue as this form grows.

File: client/src/components/Tea/TeaDetails/TeaDetails.jsx [hunk 7]
```diff
@@ -2,16 +2,17 @@
 import React from "react";
 import { Link } from "react-router-dom";
 
-export const TeaDetails = props => {
+export const TeaDetails = React.memo(props => {
   return (
     <div className="container content">
-      {props.flash ? (
+      {props.flash == "on" && (
```
Comment (JeffreyATW) [inline client/src/components/Tea/TeaDetails/TeaDetails.jsx]: Use `===` here. If you're using eslint, I'd suggest enabling the `eqeqeq` rule.

File: client/src/actions/authActions.js [hunk 8]
```diff
@@ -1,117 +1,45 @@
...
+    });
   };
 }
 
-// Set logged in user
+// Log user out
+export const logoutUser = () => dispatch => {
+  localStorage.removeItem("jwtToken");
+  setAuthToken(false);
+  dispatch({ type: "USER_LOGOUT" });
```
Comment (bdj61383) [inline client/src/actions/authActions.js:37]: You don't need to do this as part of this PR, but in general I'm a fan of having canonical types live somewhere as variables that get imported into their respective action creators and reducers.  It prevents typos like using "USER_LOGOUT" in your action creator but "USER_LOGSOUT" in your reducer, for example.

File: client/src/components/Tea/TeaEditor/TeaEditor.jsx [hunk 9]
```diff
@@ -1,62 +1,46 @@
 /* eslint-disable no-console */
 import React from "react";
 import { Link } from "react-router-dom";
-import DataList from "../../DataList";
+import { InputField } from "../../FormComponents/InputField";
 
-export const TeaEditor = props => {
+export const TeaEditor = React.memo(props => {
```
Comment (bdj61383) [inline client/src/components/Tea/TeaEditor/TeaEditor.jsx:6]: Unfortunately, this is still going to re-render each time, since the props it is passed change every time the input is changed.  You can see this by sticking a console.log or debugger statement in here.  If you want to avoid re-rendering this whole component every time TeaEditorContainer passes it new props, you'll need to make it a `React.Component` and implement your own `shouldComponentUpdate` method that ignores props that are only being passed to the `InputField`s and are not important for the `TeaEditor` parent component.

File: client/src/components/FormComponents/InputField/index.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
+import React from "react";
+
+export const InputField = React.memo(props => {
```
Comment (bdj61383) [inline client/src/components/FormComponents/InputField/index.js:3]: I'm not sure how much of a performance boost we get by trying to memoize this.  It's in the nature of inputs to change, and this will get re-rendered every time the input field is changed.  I suppose we get a boost if someone makes changes, then reverts them, etc such that we get the same inputs again, but am not sure if that offsets the processing time needed to compare the inputs every time InputField is called.  This is really an academic question though, as either case should be fast enough for us.

File: client/src/components/Tea/TeaCollectionTable/TeaCollectionTable.jsx [hunk 11]
```diff
@@ -3,149 +3,144 @@ import { Link } from "react-router-dom";
...
-              <div className="control">
-                <span className="select is-small">
-                  <select
-                    name="filterCategory"
-                    value={this.props.formControls.filterCategory}
-                    onChange={this.props.filterDropdownChangeHandler}
-                  >
-                    <option key="category" value="">
-                      Filter by
+export const TeaCollectionTable = React.memo(props => {
```
Comment (bdj61383) [inline client/src/components/Tea/TeaCollectionTable/TeaCollectionTable.jsx]: In the future, if you already have a React.Component like this (where it will always return the same output if passed the same props/state), you can just make it a React.PureComponent to get the performance boost, rather than re-writing the whole thing as a function and wrapping it in React.memo.

https://reactjs.org/docs/react-api.html#reactpurecomponent

**Review by bdj61383 — APPROVED** [review]
I'm approving this because there are no show-stoppers, but see the comments for some more things to think about.  In general I don't think there's a huge downside to wrapping your functions in React.memo, but there are caveats.  If it looks like a situation where memoization is unlikely, due to constantly changing inputs or child components that won't memoize properly, it takes (slightly) longer to check to see if there's a memoized version of a function with the given inputs and then running that function anyway when there are no matches, than it would to just run the function in the first place.
